### DISCUSSION TITLE
Amazon Basics Smart Switch support

### DISCUSSION THREAD

**Discussion by eu-gh** [arendst/Tasmota#21996] [Support request]
Hi!

I got this new(-ish?) light switch from Amazon's own generic brand, hoping to make it actually usable with Tasmota :)

Disassembly is straightforward, just a couple of Philips screws and some plastic clips.

It contains two PCBs connected by a (removable!) ribbon cable. Some pics:
![PXL_20240822_181317870](https://github.com/user-attachments/assets/f5715d4b-b25b-417f-9867-b66482289b14)
![PXL_20240822_181325323](https://github.com/user-attachments/assets/00ebc5ef-36c3-463f-8f1b-e2e644046f83)

There are absolutely zero identification points on the communication board itself, and the RF shield unfortunately seems to be soldered on. However, there are some conveniently labelled testpins on the backside, which I used to grab a serial console log from:

```ESP-ROM:esp32c3-api1-20210207
Build:Feb  7 2021
rst:0x1 (POWERON),boot:0x9 (SPI_FAST_FLASH_BOOT)
SPIWP:0xee
mode:DIO, clock div:1
Valid secure boot key blocks: 0
secure boot verification succeeded
load:0x3fcd6268,len:0x3108
load:0x403ce000,len:0x8d4
load:0x403d0000,len:0x4f58
entry 0x403ce000
I (72) boot: ESP-IDF 65e73f3b-dirty 2nd stage bootloader
I (72) boot: compile time 17:21:36
I (72) boot: chip revision: 4
I (74) boot_comm: chip revision: 4, min. bootloader chip revision: 3
I (81) boot.esp32c3: SPI Speed      : 80MHz
I (86) boot.esp32c3: SPI Mode       : DIO
I (91) boot.esp32c3: SPI Flash Size : 4MB
I (95) boot: Enabling RNG early entropy source...
I (101) boot: Partition Table:
I (104) boot: ## Label            Usage          Type ST Offset   Length
I (112) boot:  0 otadata          OTA data         01 00 0000d000 00002000
I (119) boot:  1 phy_init         RF data          01 01 0000f000 00001000
I (127) boot:  2 coredump         Unknown data     01 03 00010000 00008000
I (134) boot:  3 ota_0            OTA app          00 10 00020000 001c0000
I (142) boot:  4 ota_1            OTA app          00 11 001e0000 001c0000
I (150) boot:  5 nvs              WiFi data        01 02 003a0000 00003000
I (157) boot:  6 nvdm             WiFi data        01 02 003a3000 00003000
I (165) boot:  7 amazon           WiFi data        01 02 003a6000 0005a000
I (172) boot: End of partition table
I (176) boot: Enabled a check secure version of app for anti rollback
I (184) boot: Secure version (from eFuse) = 0
I (189) boot_comm: chip revision: 4, min. application chip revision: 3
I (196) esp_image: segment 0: paddr=00020020 vaddr=3c140020 size=638a0h (407712) map
I (266) esp_image: segment 1: paddr=000838c8 vaddr=3fc8d000 size=0675ch ( 26460) load
I (271) esp_image: segment 2: paddr=0008a02c vaddr=40380000 size=05fech ( 24556) load
I (276) esp_image: segment 3: paddr=00090020 vaddr=42000020 size=13f1a4h (1307044) map
I (476) esp_image: segment 4: paddr=001cf1cc vaddr=40385fec size=06e60h ( 28256) load
I (481) esp_image: segment 5: paddr=001d6034 vaddr=50000010 size=00014h (    20) load
I (482) esp_image: Verifying image signature...
I (487) secure_boot_v2: Verifying with RSA-PSS...
I (495) secure_boot_v2: Signature verified successfully!
I (503) boot: Loaded app from partition at offset 0x20000
I (504) secure_boot_v2: enabling secure boot v2...
I (510) secure_boot_v2: secure boot v2 is already enabled, continuing..
I (517) boot: Disabling RNG early entropy source...
I (534) cpu_start: Pro cpu up.
I (546) cpu_start: Pro cpu start user code
I (546) cpu_start: cpu freq: 160000000
I (546) cpu_start: Application information:
I (549) cpu_start: Project name:     ffo_pequod2_smart_switch
I (556) cpu_start: App version:      Pequod2B_V1.00.09-dirty
I (562) cpu_start: Secure version:   0
I (566) cpu_start: Compile time:     Mar  7 2024 09:03:26
I (572) cpu_start: ELF file SHA256:  1af2018b9de77182...
I (578) cpu_start: ESP-IDF:          65e73f3b-dirty
I (584) heap_init: Initializing. RAM available for dynamic allocation:
I (591) heap_init: At 3FCA01F0 len 0001FE10 (127 KiB): DRAM
I (597) heap_init: At 3FCC0000 len 0001F060 (124 KiB): STACK/DRAM
I (604) heap_init: At 50000030 len 00001FD0 (7 KiB): RTCRAM
I (611) spi_flash: detected chip: generic
I (615) spi_flash: flash io: dio
I (620) sleep: Configure to isolate all GPIO pins in sleep state
I (626) esp_core_dump_flash: Init core dump to flash
I (632) esp_core_dump_flash: Found partition 'coredump' @ 10000 32768 bytes
E (639) esp_core_dump_flash: Incorrect size of core dump image: 0
I (646) cpu_start: Starting scheduler.
portMAX_DELAY:ffffffff, configTICK_RATE_HZ:1000, portTICK_PERIOD_MS:1
[I/xparam_manager.c@0.657] kvdbInit:kv_user
[I/FAL] Flash Abstraction Layer (V0.5.0) initialize success.
[FlashDB] FlashDB V1.1.1 is initialize success.
[FlashDB] You can get the latest version on https://github.com/armink/FlashDB .
[I/mtask@0.680] MtaskListCreate[1]
[I/mtask@0.683] task[0]name:sys, stacksize[2048]
I (727) gpio: GPIO[7]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0 
I (736) gpio: GPIO[10]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0 
I (746) gpio: GPIO[6]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0 
I (755) gpio: GPIO[5]| InputEn: 0| OutputEn: 1| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
I (764) gpio: GPIO[4]| InputEn: 0| OutputEn: 1| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
I (794) gpio: GPIO[18]| InputEn: 0| OutputEn: 1| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
I (795) gpio: GPIO[19]| InputEn: 0| OutputEn: 1| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
[V/xparam_manager.c@0.773] param id:[5] change notify
[D/task_sys@0.774] parameter: event_id:102, code:1, msg_data:5
[D/lds_smart_switch.c@0.776] the device isn't registered power on!
[D/lds_smart_switch.c@0.783] count log click times,count [1]
esp_dram_heap_stats registered
Unable to register esp_iram_heap_stats
[1970-01-01 00:00:00.261 W   0 4208d2ca]aflws_create_context: LWS: 4.3.0, NET CLI H1 MQTT SS-STATIC-POL MbedTLS IPv6-absent
w:KVS Partition kvsBackup not i mem: platform fd map:   120 bytes
nit, may result in loss of data
Returning default provider: esp_dram_heap_s 
[1970-01-01 00:00:00.806 W   0 00000000]Warning: 2 log lines dropped
[D/ffo_main.c@2.795] Smart Home Application pre-initialization...
-->Enter smart home app ...
[V/ack_core_lifecycle.c@2.818] ACK Device SDK Implementation Core (re-)started; firmware version 0x202308191001
[I/ack_core_lifecycle.c@2.825] Resetting ACK module.
[I/ack_hal_implement.c@3.081] ACK lifecycle state: UNAVAILABLE
[I/ack_core_lifecycle.c@3.084] ACK Device SDK Implementation Core fully initialized.
Initializing OTA callback logic
Start smart home app loop...
[I/ack_core_processevent.c@3.098] HOSTINT asserted; processing inbound event.
[I/ack_core_processevent.c@3.105] Received connectivity module booted event.
[I/ack_core_lifecycle.c@3.113] Connectivity Module finished booting or rebooting.
[I/ack_hal_implement.c@3.122] ACK lifecycle state: BOOTED
[D/lds_smart_switch.c@3.128] the power status (0) save to flash
[V/xparam_manager.c@3.137] param id:[2] change notify
[D/task_sys@3.14cycle.c@3.142] Sending host config to ACK Connectivi[I/ack_core_lifecycle.c@3.142] Sending host config to ACK Connectivity Module...
[I/ack_core_lifecycle.c@3.156] Sent host config to ACK Connectivity Module.
[I/ack_core_lifecycle.c@3.164] Module diagnostic device type: A11CN8MUNVD0FZ
[I/ack_core_lifecycle.c@3.172] Module diagnostic firmware version: protocol 1070222302, build 1070130760, incremental version A11CN8MUNVD0FZ
[I/ack_core_lifecycle.c@3.186] Module DSN: 
[I/ack_core_lifecycle.c@3.190] Module provisioning state: 5
[D/lds_smart_switch.c@3.198] ACK Status: 1(ack not registered)
[I/ack_hal_implement.c@3.204] ACK lifecycle state: MODULE NOT REGISTERED
[D/device_line_detect.c@3.310] ldsDeviceLineDetectInit
[D/lds_smart_switch.c@3.315] the ACK info -> DSN : A13SS1F8OSJ1YATZID, TYPE : A11CN8MUNVD0FZ,VERSION : 202403071
[D/lds_smart_switch.c@3.324] loop:ProvisonDoneGet = 0,ack_connect_state = 1
[D/lds_smart_switch.c@5.731] ACK Status: 1(ack not registered)
[V/xparam_manager.c@5.777] param id:[5] change notify
[D/task_sys@5.781] parameter: event_id:102, code:1, msg_data:5
```

It appears to be an ESP-based chip, which is awesome! Can anybody assist me on identifying the pinouts so I can put this thing into flash mode?

Thank you and kind regards!

**Comment by s-hadinger** [comment]
Indeed. It seems also that the esp32c3 is not locked by eFuse. So flashing should be possible. 

Markings are clear for Tx and Rx. It's possible that flashing gpio is on one of the buttons. 

Do you have a link on Amazon?

**Comment by eu-gh** [comment]
Any idea what I can try?

Link is https://www.amazon.de/dp/B0CRKTXMF7

**Comment by sfromis** [comment]
When you have it flashed, I'd expect it to be feasible to find the buttons and relay using the methods described on this page:
https://tasmota.github.io/docs/Configuration-Procedure-for-New-Devices/
Note that you cannot directly use the templates there for ESP32-C3, but the concept of assigning pins as Relay or Switch, and then see what happens should work well.

**Comment by s-hadinger** [comment]
Try rebooting with the buttons pressed and see in serial logs if you are in Flash mode

**Comment by Jason2866** [comment]
@s-hadinger I would not make a bet that the C3 can be flashed. Secure boot is enabled.

**Reply by s-hadinger** [reply]
Well I don't have a reference boot at hand. The following line made me thought that secure boot was not indeed effective 

`Valid secure boot key blocks: 0`

**Comment by sfromis** [comment]
The ESP32-C3 module looks a lot like the one described here (LA66701 from Leedarson Lighting), including pinout.
https://device.report/manual/9346936
Of course, I can't say anything for sure, but the similarities are striking.

In that case, the pin (gpio 9) to be grounded for boot is in the middle of the strip, but not on the top, instead on the part that is soldered to the main low-voltage PCB.

My guess would be that this is the pin connected to the R2 resistor a bit to the right. This may be a pull-up for one of the buttons, maybe SW1. You could also try grounding it at the bottom end of R2, as this *looks* like a trace going to the center pin.

**Comment by eu-gh** [comment]
Thank you! The LA66701 seems to be an exact match.

However, I still can't seem to access the flash via esptool. According to this layout:
![grafik](https://github.com/user-attachments/assets/bac069e0-aa58-449f-aa39-800b6631e450)
Pin 5 is the first GPIO (they call it "IO1", but normal arrays starts at zero, so should be GPIO0) and is connected to R3, which gets pulled to GND by pressing SW1 (the regular user "reset" button). Following this logic, pin 9 (the centered one, IO3) should be GPIO2. In addition to pressing the button at boot, I tried pulling this to GND too, but it doesn't seem to change anything, the bootlog also is the same.

Also, I found this: https://community.home-assistant.io/t/help-trying-to-flash-esp32-in-home-depot-bulb/326294 which isn't the same module, but made by the same company, apparently they weren't successful either.

I wonder whether we need a different pin in this case? For example 10 and 15 are documented as "download control" (with a typo).

BTW, secure boot on a microcontroller? What a time to be alive!

**Reply by sfromis** [reply]
Concepts like "normal arrays" are not relevant here, meaning that I'd not expect IO1 to refer to gpio0. And it looks like I was mistaken about the center pin, as the numbers are simply physical numbers on the module, where you found the relationship in the section "Pin Definition":
![image](https://github.com/user-attachments/assets/4d110068-f6a6-49e2-b701-56de15a01a5a)

Cross-referencing to ESP32-C3 chip pinout, it looks like the important gpio 9 is on chip pin 15, which the above table coincidentally maps as module pin 15, 2nd from the right on the back side, or 2nd from the left seen from the top side (with antenna edge up). For some reason, they indeed chose to call this the "PL" pin. I see no clues for what this pin may be connected to on the PCB. No particular reason to assume this to be one of the 3 physical buttons.

Still, there is a very real risk that the doubt about flashability expressed by Jason2866 is well-founded, leaving this pin hunt a bit of an academic exercise.

**Reply by eu-gh** [reply]
So I'm getting the feeling that de-soldering the daughterboard to access the flash pins seems like a waste of time, since there is no known way of bypassing the secure boot feature?

**Reply by Jason2866** [reply]
**If** secure boot is active the only way is replacing the module.

**Reply by 1RandomDev** [reply]
I'd probably just replace the chip inside with the one from an ESP32 Super Mini, since the module has a very specific pinout and shape.

**Comment by barbudor** [comment]
Interresting device
- is neutral required ? I would assume yes seeing the number of screw connections
- is it a rocker switch or a push button ?

Thanks

**Reply by eu-gh** [reply]
Yes, there are screw terminals for line, neutral, switched power and a blank one. Its cover has a spring loaded rocker. which means that the on/off state buttons can be pressed at any time, but not simultaneously.

**Comment by 1RandomDev** [comment]
I have the European variant of the Amazon Basics smart socket which uses the same WiFI module and as described in this [Reddit post](https://www.reddit.com/r/esp32/comments/1v9dgcf/comment/p0cyuoa/) I made there is no way to program it without replacing the chip.
